pkg/util/admission/granter.go

```diff
@@ -2106,9 +2199,186 @@ func (*ioLoadListener) adjustTokensInner(
+	}
+	// Compute flush tokens for this interval that would cause 100% utilization.
+	intFlushTokens := float64(im.Flush.WriteThroughput.PeakRate()) * adjustmentInterval
+	intWriteStalls := cumWriteStallCount - prev.cumWriteStallCount
+
+	// Ensure flushUtilTargetFraction is in the configured bounds. This also
+	// does lazy initialization.
+	const maxFlushUtilTargetFraction = 1.5
+	flushUtilTargetFraction := prev.flushUtilTargetFraction
+	if flushUtilTargetFraction == 0 {
+		// Initialization: use the maximum configured fraction.
+		flushUtilTargetFraction = minFlushUtilTargetFraction
+		if flushUtilTargetFraction < maxFlushUtilTargetFraction {
+			flushUtilTargetFraction = maxFlushUtilTargetFraction
+		}
+	} else if flushUtilTargetFraction < minFlushUtilTargetFraction {
+		// The min can be changed in a running system, so we bump up to conform to
+		// the min.
+		flushUtilTargetFraction = minFlushUtilTargetFraction
+	}
+	numFlushTokens := int64(unlimitedTokens)
+	// doLogFlush becomes true if something interesting is done here.
+	doLogFlush := false
+	smoothedNumFlushTokens := prev.smoothedNumFlushTokens
+	const flushUtilIgnoreThreshold = 0.05
+	if intFlushUtilization > flushUtilIgnoreThreshold {
+		if smoothedNumFlushTokens == 0 {
+			// Initialization.
+			smoothedNumFlushTokens = intFlushTokens
+		} else {
+			smoothedNumFlushTokens = alpha*intFlushTokens + (1-alpha)*prev.smoothedNumFlushTokens
+		}
+		const flushUtilTargetFractionIncrement = 0.025
+		// Have we used, over the last (15s) cycle, more than 90% of the tokens we
+		// would give out for the next cycle? If yes, highTokenUsage is true.
+		highTokenUsage :=
+			float64(prev.tokensUsed) >= 0.9*smoothedNumFlushTokens*flushUtilTargetFraction
+		if intWriteStalls > 0 {
+			// Try decrease since there were write-stalls.
+			numDecreaseSteps := 1
+			// These constants of 5, 3, 2, 2 were found to work reasonably well,
+			// without causing large decreases. We need better benchmarking to tune
+			// such constants.
+			if intWriteStalls >= 5 {
+				numDecreaseSteps = 3
+			} else if intWriteStalls >= 2 {
+				numDecreaseSteps = 2
+			}
+			for i := 0; i < numDecreaseSteps; i++ {
+				if flushUtilTargetFraction >= minFlushUtilTargetFraction+flushUtilTargetFractionIncrement {
+					flushUtilTargetFraction -= flushUtilTargetFractionIncrement
+					doLogFlush = true
+				} else {
+					break
+				}
+			}
+		} else if flushUtilTargetFraction < maxFlushUtilTargetFraction-flushUtilTargetFractionIncrement &&
+			intWriteStalls == 0 && highTokenUsage {
+			// No write-stalls, and token usage was high, so give out more tokens.
+			flushUtilTargetFraction += flushUtilTargetFractionIncrement
+			doLogFlush = true
+		}
+		if highTokenUsage {
+			doLogFlush = true
+		}
+		flushTokensFloat := flushUtilTargetFraction * smoothedNumFlushTokens
+		if flushTokensFloat < float64(math.MaxInt64) {
+			numFlushTokens = int64(flushTokensFloat)
```

Comment:
Similar to the number of available CPU slots, should we export the number of flush tokens and compaction byte tokens as metrics?